### PR TITLE
Add ability to index graph depth.

### PR DESCRIPTION
**Pull request by josephwillard**
This gives the user an option to provide either a tuple, list or range which is used to limit the output of the graph. In the following example,

```
import pymc4 as pm

from pymc4 import distributions as dist

from tensorflow.python.eager.context import graph_mode  

from symbolic_pymc.tensorflow.printing import tf_dprint


@pm.model
def transform_example():
    x = yield dist.Normal('x', mu=0, sigma=1)
    y = yield dist.Normal('y', mu=0, sigma=1)
    q = tf.realdiv(x, y)
    return q


with graph_mode():
    model = transform_example()
    obs_graph, state = pm.evaluate_model(model)

tf_dprint(obs_graph, index=range(3, 6))
```

the output is,

```
In [18]: 
|  |  |  Op(Reshape)	"Normal_11_1/sample/Reshape"
|  |  |  |  Tensor(Add):0,	shape=[1]	"Normal_11_1/sample/add:0"
|  |  |  |  |  Op(Add)	"Normal_11_1/sample/add"
|  |  |  |  Tensor(ConcatV2):0,	shape=[0]	"Normal_11_1/sample/concat_1:0"
|  |  |  |  |  Op(ConcatV2)	"Normal_11_1/sample/concat_1"
|  |  |  Op(Reshape)	"Normal_12_1/sample/Reshape"
|  |  |  |  Tensor(Add):0,	shape=[1]	"Normal_12_1/sample/add:0"
|  |  |  |  |  Op(Add)	"Normal_12_1/sample/add"
|  |  |  |  Tensor(ConcatV2):0,	shape=[0]	"Normal_12_1/sample/concat_1:0"
|  |  |  |  |  Op(ConcatV2)	"Normal_12_1/sample/concat_1"
```

### CODE REVIEW

File: symbolic_pymc/tensorflow/printing.py [hunk 1]
```diff
@@ -58,41 +67,41 @@ def tf_dprint(obj, printer=None):
...
 
-    _tf_dprint(obj, printer)
+    _tf_dprint(obj, printer, depth=0)
 
 
 @singledispatch
-def _tf_dprint(obj, printer):
-    printer.println(obj)
+def _tf_dprint(obj, printer, depth):
+    if printer.depth_check(depth):
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Why have the checking logic in the `TFlowPrinter` object when, especially when the depth value is exclusively managed *outside* the `TFlowPrinter` object?

You could probably track depth in the same way/alongside as the indentation level, or&mdash;better yet&mdash;you could change the indentation tracking to depth tracking (if only in naming) and reasonably double-purpose that.

File: symbolic_pymc/tensorflow/printing.py [hunk 2]
```diff
@@ -41,8 +42,16 @@ def println(self, obj):
...
+        if isinstance(self.depth_index, (list, tuple, range)):
+            return current_depth in self.depth_index
+        elif self.depth_index is None:
+            return True
+        else:
+            raise ValueError("Depth needs to be a tuple, list or range type.")
+
 
-def tf_dprint(obj, printer=None):
+def tf_dprint(obj, printer=None, index=None):
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: This needs a good explanation of what `index` is/does, especially when it's a collection.

File: symbolic_pymc/tensorflow/printing.py (hunk 1 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: I took your advice and renamed `self.indentation` to `self.depth_tracking`, the functionality is still the same and depth is now calculated based off this.

File: symbolic_pymc/tensorflow/printing.py (hunk 2 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: I renamed `index` to `depth_index` to make it more clear what it does.

File: symbolic_pymc/tensorflow/printing.py [hunk 3]
```diff
@@ -16,23 +16,23 @@ def __init__(self, formatter, buffer, depth_index):
         self.buffer = buffer
         self.formatter = formatter
         self.depth_index = depth_index
-        self.indentation = ""
+        self.depth_tracking = ""
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: I don't understand this renaming.  Now, it's much less clear what that variable is for/doing based on its name.

File: symbolic_pymc/tensorflow/printing.py [hunk 4]
```diff
@@ -42,19 +42,19 @@ def println(self, obj):
         self.buffer.write(self.format(obj) + "\n")
         self.buffer.flush()
 
-    def depth_check(self, current_depth):
+    def depth_check(self):
         if isinstance(self.depth_index, (list, tuple, range)):
-            return current_depth in self.depth_index
+            return self.depth_tracking.count("|") in self.depth_index
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Track the depth separately as a count; this approach is unnecessarily format-dependent and computationally wasteful.  

Plus, if you do it that way, it would be much easier to change the output so that a subgraph can be displayed as if it were the root of the graph.

Comment (brandonwillard) [reply symbolic_pymc/tensorflow/printing.py]: So the only way to specify a graph depth lower and upper bound is to use `range(0, upper_bound)` and `range(lower_bound, sys.maxsize)`?  That doesn't sound as simple as having, say, only upper and lower bound depth parameters.

File: symbolic_pymc/tensorflow/printing.py (hunk 3 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: TFlowPrinter is now setup to track depth separately. The original naming convention was returned.

File: symbolic_pymc/tensorflow/printing.py (hunk 4 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: When specifying graph depth one doesn't have to use `range`, there is also the option to use a `list` or `tuple` object. I updated the function with a docstring with an example.

File: symbolic_pymc/tensorflow/printing.py [hunk 5]
```diff
@@ -41,11 +47,50 @@ def println(self, obj):
...
+            raise ValueError("Depth needs to be a tuple, list or range type.")
+
 
-def tf_dprint(obj, printer=None):
+def tf_dprint(obj, printer=None, depth_index=None):
     """Print a textual representation of a TF graph.
 
     The output roughly follows the format of `theano.printing.debugprint`.
+
+    with graph_mode():
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: If you're going to add multi-line code examples in a docstring, we'll need to determine the proper formatting (e.g. RST for Sphinx rendering) and apply it universally.  When in doubt, follow PyMC3's conventions.

File: symbolic_pymc/tensorflow/printing.py [hunk 6]
```diff
@@ -41,11 +47,50 @@ def println(self, obj):
...
+        t1 = [[[1, 2], [2, 3]], [[4, 4], [5, 3]]]
+        t2 = [[[7, 4], [8, 4]], [[2, 10], [15, 11]]]
+        t3 = tf.concat([t1, t2], -1)
+
+    In [1]: tf_dprint(t3)
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Use standard Python console formatting (i.e. `>>>`).  I believe there's an IPython magic for that.

File: symbolic_pymc/tensorflow/printing.py [hunk 7]
```diff
@@ -12,23 +12,29 @@
...
+        self.depth_index = depth_index
+        self.depth_count = 0
         self.indentation = ""
         self.printed_subgraphs = set()
 
     @contextmanager
     def indented(self, indent):
         pre_indentation = self.indentation
+        pre_depth_count = self.depth_count
         if isinstance(indent, int):
```
[the remark this answers is not in the document]
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: `self.depth_count` was removed and set to increment anytime the method is called.

[the remark this answers is not in the document]
Comment (brandonwillard) [reply symbolic_pymc/tensorflow/printing.py]: You're still have logic pertaining to `indent` being an integer, which it shouldn't be.

File: tests/tensorflow/test_printing.py [hunk 8]
```diff
@@ -94,16 +94,26 @@ def test_shape_index():
...
     |  |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
     |  |  |  |  |  Op(Mul)\t"y"
+    .
+    .
+    .
     |  |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
     |  |  |  |  |  Op(Mul)\t"y"
+    .
+    .
+    .
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: The purpose of this indexing functionality is to make large graphs viewable&mdash;in part by restricting the output to only certain subgraphs&mdash;but this new form of output doesn't accomplish that.  

Before this change, a large graph would push all the relevant content to the right an amount of pipes equal to the graph's lower-indexed depth, and now, with this change, it will **also** flood the screen vertically.  The end result is a negligible improvement over the non-indexed output.

More helpful truncated/indexed output would show **only the indexed subgraph** (i.e. without any of the preceding depth indicators/pipes) and a simple indication of the depth(s) of said subgraph.

For example:
```
3 <= depth <= 5
... Op(Add)  "x_p_y"
... |  Tensor(Mul):0,  shape=[None, 1]  "y:0"
... |  |  Op(Mul)  "y"
... |  |  | ...
... |  Tensor(Mul):0,  shape=[None, 1]  "z:0"
... |  |  Op(Mul)  "z"
... |  |  | ...
```

Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: I removed the original notation for truncating and made it resemble the example above.

File: tests/tensorflow/test_printing.py [hunk 9]
```diff
@@ -94,51 +94,34 @@ def test_shape_index():
...
-    |  |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
-    .
-    .
-    .
-    |  |  |  |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
-    |  |  |  |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
+    ... Op(Add)\t"x_p_y"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
+    ... |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: You're missing an indicator for the ommitted depths here.

File: symbolic_pymc/tensorflow/printing.py [hunk 10]
```diff
@@ -38,14 +42,62 @@ def print(self, obj):
...
+    
+    Parameters
+    ----------
+    obj : Tensorflow object
+        Tensorflow graph object to be represented.
+    printer : optional
+        Backend used to display the output.
+    depth_index : list, tuple or range
+        Used to index specific portions of the graph.
     """
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Why are you allowing non-contiguous depths?

Regardless, as I mentioned before, the interface would be much clearer and easier to use if there were optional depth upper- and lower-bound parameters only.

File: symbolic_pymc/tensorflow/printing.py [hunk 11]
```diff
@@ -38,14 +42,62 @@ def print(self, obj):
...
+        >>> tf_dprint(t3)
+        Tensor(ConcatV2):0,	shape=[2, 2, 4]	"concat_2:0"
+        |  Op(ConcatV2)	"concat_2"
+        |  |  Tensor(Const):0,	shape=[2, 2, 2]	"concat_2/values_0:0"
+        |  |  Tensor(Const):0,	shape=[2, 2, 2]	"concat_2/values_1:0"
+        |  |  Tensor(Const):0,	shape=[]	"concat_2/axis:0"
+
+
+        >>> tf_dprint(t3, depth_index=[2])
+        |  |  Tensor(Const):0,	shape=[2, 2, 2]	"concat_2/values_0:0"
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: This output is no longer valid.

Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: Fixed in latest commit.

File: symbolic_pymc/tensorflow/printing.py (hunk 10 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: Took your advice and added `depth_lower` and `depth_upper` bound parameters.

File: tests/tensorflow/test_printing.py (hunk 9 above)
Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Fixed in latest commit.

File: symbolic_pymc/tensorflow/printing.py [hunk 12]
```diff
@@ -38,14 +42,58 @@ def print(self, obj):
         self.buffer.flush()
 
     def println(self, obj):
-        self.buffer.write(self.format(obj) + "\n")
-        self.buffer.flush()
+        try:
+            if None not in self.depth_index and self.depth_count in range(
+                self.depth_index[0], self.depth_index[1]
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Why not pass a `range`, or at least convert to one, in the constructor?

File: tests/tensorflow/test_printing.py [hunk 13]
```diff
@@ -79,3 +79,67 @@ def test_unknown_shape():
...
+
+    A = tf.compat.v1.placeholder('float', name='A',
+                                 shape=tf.TensorShape([None, None]))
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+    tf_dprint(z, depth_lower=3.2, depth_upper=6.2)
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: This isn't a real test.

File: tests/tensorflow/test_printing.py [hunk 14]
```diff
@@ -79,3 +79,67 @@ def test_unknown_shape():
...
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+
+    range_out = io.StringIO()
+    with redirect_stdout(range_out):        
+        tf_dprint(z, depth_lower=3, depth_upper=6)
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: Use a graph and/or range that truncates a valid cross-section/slice of the graph (i.e. doesn't go past the beginning or end of a graph).

File: symbolic_pymc/tensorflow/printing.py (hunk 12 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: I choose to convert to a range in my latest commit.

File: tests/tensorflow/test_printing.py (hunk 13 above)
Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Updated in my latest commit, let me know if it looks better.

File: tests/tensorflow/test_printing.py (hunk 14 above)
Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Updated in latest commit.

File: tests/tensorflow/test_printing.py [hunk 15]
```diff
@@ -79,3 +79,72 @@ def test_unknown_shape():
...
+    ... |  |  Op(Mul)\t"y"
+    ... |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
+    ... |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
+    ''')
+
+    assert range_out.getvalue() == expected_out.lstrip()
+
+    range_out = io.StringIO()
+    with redirect_stdout(range_out):        
+        tf_dprint(z)
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: This looks like a copy-paste of one of the older tests above.  Did you mean to set `depth_lower` and `depth_upper`?  

This might be a good opportunity to test a *real* cross-section of graph (e.g. maybe `depth_lower=1`, `depth_upper=4`)&mdash;in other words, something that truncates above, below, and across more than one branch at the same level.

Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Updated using the suggested depths. I also removed the second portion of the test since it gets used in a separate test.

File: symbolic_pymc/tensorflow/printing.py [hunk 16]
```diff
@@ -38,14 +47,53 @@ def print(self, obj):
         self.buffer.flush()
 
     def println(self, obj):
-        self.buffer.write(self.format(obj) + "\n")
-        self.buffer.flush()
+        if self.depth_index and self.depth_count in self.depth_index:
+            self.buffer.write(
+                self.format(obj).replace("|  " * min(self.depth_index), "... ", 1) + "\n"
```
[the remark this answers is not in the document]
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: In the latest commit I moved the `replace` statement into the indented method.

File: symbolic_pymc/tensorflow/printing.py [hunk 17]
```diff
@@ -38,14 +47,53 @@ def print(self, obj):
...
+            )
+            self.buffer.flush()
+        elif self.depth_index is None:
+            self.buffer.write(self.format(obj) + "\n")
+            self.buffer.flush()
+        else:
```
[the remark this answers is not in the document]
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: Removed the unnecessary `else` statement in the latest commit.

File: symbolic_pymc/tensorflow/printing.py [hunk 18]
```diff
@@ -12,23 +12,36 @@
...
     @contextmanager
     def indented(self, indent):
         pre_indentation = self.indentation
+        self.depth_count += 1
         if isinstance(indent, int):
             self.indentation += " " * indent
         else:
             self.indentation += indent
+
+        if self.depth_index and "... " not in self.indentation:
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Why do you need to search and replace in the strings you're responsible for creating?!  Why can't you explicitly craft the correct value of `self.indentation` in the first place?

File: tests/tensorflow/test_printing.py [hunk 19]
```diff
@@ -79,3 +79,47 @@ def test_unknown_shape():
...
+                                 shape=tf.TensorShape([None, None]))
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+    range_out = io.StringIO()
+    with redirect_stdout(range_out):        
+        tf_dprint(z, depth_lower=1, depth_upper=4)
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: Try `depth_upper >= 5`; we want a test that has upper-bound trucation across more than one term that has 
a non-trivial subgraph (i.e. more than one "..." at the same depth).

File: symbolic_pymc/tensorflow/printing.py (hunk 18 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: I removed the search in the latest commit.

File: tests/tensorflow/test_printing.py (hunk 19 above)
Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Updated to use `depth_upper = 5`.

File: symbolic_pymc/tensorflow/printing.py [hunk 20]
```diff
@@ -81,11 +127,11 @@ def _(obj, printer):
     _tf_dprint(prefix, printer)
     if len(obj.op.inputs) > 0:
         with printer.indented("|  "):
-            if obj not in printer.printed_subgraphs:
+            if obj in printer.printed_subgraphs and printer.depth_index is None:
```
Comment (brandonwillard) [inline symbolic_pymc/tensorflow/printing.py]: Is this disabling the do-not-reprint-already-printed-subgraphs logic when limits are specified?  If so, that shouldn't happen.  

If the `printer.depth_index` condition is there because subgraphs that occur before the desired range are being added to `printer.printed_subgraphs` and, as a result, don't get printed in the desired depth range (i.e. when that range is at a greater depth), then that's a clear sign that this logic isn't correct.

It might make much more sense to put this do-not-reprint-already-printed-subgraphs logic in `TFlowPrinter.println` and modulate in accordance with `TFlowPrinter.depth_index` and `TFlowPrinter.depth_count` when subgraphs/terms are added to `printer.printed_subgraphs`&mdash;in the same way you could directly and concisely determine `TFlowPrinter.indentation` in `TFlowPrinter.indented`.

No matter what, this PR needs a test for depth limits + repeated graph elements.

File: tests/tensorflow/test_printing.py [hunk 21]
```diff
@@ -79,3 +79,47 @@ def test_unknown_shape():
...
+
+    expected_out = textwrap.dedent('''
+    ... Op(MatMul)\t"A_dot"
+    ... |  Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... |  |  Op(Add)\t"x_p_y"
+    ''')
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: Shouldn't this end with a `... |  |  |  ...` to indicate the truncated subgraph?

Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Added in latest commit.

File: symbolic_pymc/tensorflow/printing.py (hunk 20 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: In the latest commit this no longer disables the do-not-reprint-already-printed-subgraphs logic when limits are specified.

Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: In the latest commit I added a new method that adds to `printed_subgraphs`. Adding do-not-reprint-already-printed-subgraphs  logic to `TFlowPrinter.println` runs into an issue since both `tensors` and `op` objects get passed.

File: tests/tensorflow/test_printing.py [hunk 22]
```diff
@@ -79,3 +79,51 @@ def test_unknown_shape():
...
+    range_out = io.StringIO()
+    with redirect_stdout(range_out):        
+        tf_dprint(z, depth_lower=1, depth_upper=5)
+
+    expected_out = textwrap.dedent('''
+    ... Op(MatMul)\t"A_dot"
+    ... |  Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... |  |  Op(Add)\t"x_p_y"
+    ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
```
[the remark this answers is not in the document]
Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: In the recent commit I added a trailing `...` to indicate the subgraph is being truncated.

File: symbolic_pymc/tensorflow/printing.py (hunk 20 above)
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: Added tests for both in the latest commit.

File: symbolic_pymc/tensorflow/printing.py [hunk 23]
```diff
@@ -12,40 +12,102 @@
...
         try:
             yield
         finally:
             self.indentation = pre_indentation
+            self.depth_count -= 1
 
     def format(self, obj):
-        return self.indentation + self.formatter(obj)
+        if self.depth_index and obj is None:
+            return "... " + "|  " * max(self.depth_index) + "..."
```
[the remark this answers is not in the document]
Comment (josephwillard) [reply symbolic_pymc/tensorflow/printing.py]: Moved some things around, let me know how they look. I also added a check to determine if there is a lower-bound truncation.

File: tests/tensorflow/test_printing.py [hunk 24]
```diff
@@ -104,13 +104,14 @@ def test_shape_index():
     ... |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
     ... |  |  Op(Add)\t"x_p_y"
     ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  ...
     ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
-    ... |  |  |  |  ...
+    ... |  |  |  ...
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: The ellipses indicate that everything above/below the level they're on is truncated.  This output says (twice) that the level those `Tensor(Mul)`s are on is being truncated.  
In other words, it looks like you need to (re)add a pipe before those ellipses, no?

Comment (josephwillard) [reply tests/tensorflow/test_printing.py]: Added the pipe back so the levels should make more sense now.

File: tests/tensorflow/test_printing.py [hunk 25]
```diff
@@ -79,3 +79,90 @@ def test_unknown_shape():
...
+    ... Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... Op(Add)\t"x_p_y"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  Op(Mul)\t"y"
+    ... |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
+    ... |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  ...
+    ... ...
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: The expected output should be as follows, no?
```
... Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
... Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
... |  Op(Add)\t"x_p_y"
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  Op(Mul)\t"y"
... |  |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
... |  |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  ...
```

File: tests/tensorflow/test_printing.py [hunk 26]
```diff
@@ -79,3 +79,90 @@ def test_unknown_shape():
...
+    ... Tensor(MatMul):0,\tshape=[None, 1]\t"A_dot:0"
+    ... Op(MatMul)\t"A_dot"
+    ... |  Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... |  |  Op(Add)\t"x_p_y"
+    ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  |  ...
+    ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  |  ...
+    ... ...
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: I believe the expected output should be as follows:
```
Tensor(MatMul):0,\tshape=[None, 1]\t"A_dot:0"
|  Op(MatMul)\t"A_dot"
|  |  Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
|  |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
|  |  |  Op(Add)\t"x_p_y"
|  |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
|  |  |  |  |  ...
|  |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
|  |  |  |  |  ...
```

File: tests/tensorflow/test_printing.py [hunk 27]
```diff
@@ -79,3 +79,90 @@ def test_unknown_shape():
...
+    ... Op(Add)\t"x_p_y"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  Op(Mul)\t"y"
+    ... |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
+    ... |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  ...
+    ... ...
+    ''')
+
```
Comment (brandonwillard) [inline tests/tensorflow/test_printing.py]: This PR has a test that starts from a lower-bound that's at the minimum depth and an upper-bound that's before the maximum, as well as a test with a lower-bound greater than the minimum and an upper-bound at (or greater than) the maximum depth.
Now, all it needs is a test for a lower-bound above the minimum and an upper-bound below the maximum.

For example, `tf_dprint(z, depth_lower=3, depth_upper=5)`, which should have the following expected output:
```
... Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
... Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
... |  Op(Add)\t"x_p_y"
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  ...
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  ...
```
